### PR TITLE
Fix incorrect checks in `io.ballerina.runtime.internal.types.BFutureType#equals`

### LINKED ISSUE #30280: `io.ballerina.runtime.internal.types.BFutureType#equals` is incorrect
**Description:**
Checks if the other type is a table type instead of a future type at https://github.com/ballerina-platform/ballerina-lang/blob/vswan-lake-alpha5/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BFutureType.java#L72

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #30280

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type_cast_expr.bal [hunk 1]
```diff
@@ -377,6 +377,13 @@ function testFutureCastPositive() returns boolean {
     return s1 === s2;
 }
 
+function testFutureWithoutFutureConstraintCastPositive() {
+    future f1 = start getNewEmployee();
+    any a = f1;
+    future f2 = <future> a;
+    test:assertTrue(f1 === f2);
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type_cast_expr.bal:384]: Lets add test with false case.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal [hunk 2]
```diff
@@ -60,6 +60,12 @@ public function testAlternateWaitTwiceOnTwoFutures() {
     assertError(r1);
 }
 
+function testFutureEqualityNegative() {
+    future<int> f1 = start calculate(2, 3);
+    future f2 = start calculate(2, 3);
+    test:assertFalse(f1 === f2);
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal:66]: IMO reference equal check will not cover the above code. Shall we remove this test?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BFutureType.java [hunk 3]
```diff
@@ -69,7 +69,12 @@ public int getTag() {
 
     @Override
     public boolean equals(Object obj) {
-        if (!super.equals(obj) || !(obj instanceof BTableType)) {
+
+        if (!(obj instanceof BFutureType)) {
+            return false;
+        }
+
+        if (!super.equals(obj)) {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BFutureType.java:77]: I think  if (!super.equals(obj))  check does not need. we can do same as BParameterizedType

Comment (warunalakshitha) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BFutureType.java:77]: ex.        
```
        if (this == obj) {
            return true;
        }

        if (!(obj instanceof BFutureType)) {
            return false;
        }
```

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type_cast_expr.bal (hunk 1 above)
[the remark this answers is not in the document]
Comment (warunalakshitha) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type_cast_expr.bal:384]: ex.

```
function testFutureWithoutFutureConstraintCastPositive() {
    future f1 = start getNewEmployee();
    any a = f1;
    future f2 = <future> a;
```

[the remark this answers is not in the document]
Comment (gabilang) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type_cast_expr.bal]: Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/31605/commits/79a29a366dfdb5fc3c8af63087cd849d13560816

File: tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal [hunk 4]
```diff
@@ -60,6 +60,12 @@ public function testAlternateWaitTwiceOnTwoFutures() {
     assertError(r1);
 }
 
+function testFutureEqualityNegative() {
```
[the remark this answers is not in the document]
Comment (gabilang) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/future/future_negative.bal]: Removed.

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**